test/callbacks.jl

```diff
@@ -0,0 +1,208 @@
+    @test id.input[1] == (:id1, :prop1)
+    @test id.output[1] == (:id2, :prop2)
+
+    id = callid"{state1.prop1, state2.prop2} input1.prop1, input2.prop2 => output1.prop1, output2.prop2"
+    @test id isa CallbackId
+    @test length(id.state) == 2
+    @test length(id.input) == 2
+    @test length(id.output) == 2
+    @test id.input[1] == (:input1, :prop1)
+    @test id.input[2] == (:input2, :prop2)
+    @test id.output[1] == (:output1, :prop1)
+    @test id.output[2] == (:output2, :prop2)
+    @test id.state[1] == (:state1, :prop1)
+    @test id.state[2] == (:state2, :prop2)
+end
+
+@testset "callback!" begin
+    app = dash()
+    app.layout = html_div() do
+            dcc_input(id = "my-id", value="initial value", type = "text"),
+            html_div(id = "my-div")        
+        end
+
+    callback!(app, callid"my-id.value => my-div.children") do value
+        return value
+    end
+    @test length(app.callbacks) == 1
+    @test haskey(app.callbacks, Symbol("my-div.children"))
+    @test app.callbacks[Symbol("my-div.children")].func("test") == "test"
+
+    handler = make_handler(app)
+    request = HTTP.Request("GET", "/_dash-dependencies")
+    resp = HTTP.handle(handler, request)
+    deps = JSON2.read(String(resp.body))
+
+    @test length(deps) == 1
+    cb = deps[1]
+    @test cb.output == "my-div.children"
+    @test cb.inputs[1].id == "my-id"
+    @test cb.inputs[1].property == "value"
+    @test :clientside_function in keys(cb)
+    @test isnothing(cb.clientside_function)
+
+    app = dash()
+    app.layout = html_div() do
+            dcc_input(id = "my-id", value="initial value", type = "text"),
+            html_div(id = "my-div"),
+            html_div(id = "my-div2")    
+        end
+    callback!(app, callid"{my-id.type} my-id.value => my-div.children, my-div2.children") do state, value
+        return state, value
+    end
+    @test length(app.callbacks) == 1
+    @test haskey(app.callbacks, Symbol("..my-div.children...my-div2.children.."))
+    @test app.callbacks[Symbol("..my-div.children...my-div2.children..")].func("state", "value") == ("state", "value")
+
+    app = dash()
+     
+    app.layout = html_div() do
+            dcc_input(id = "my-id", value="initial value", type = "text"),
+            html_div(id = "my-div"),
+            html_div(id = "my-div2")    
+        end
+    
+    callback!(app, callid"my-id.value => my-div.children") do value
+        return value
+    end
+    callback!(app, callid"my-id.value => my-div2.children") do value
+        return "v_$(value)"
+    end
+
+    @test length(app.callbacks) == 2
+    @test haskey(app.callbacks, Symbol("my-div.children"))
+    @test haskey(app.callbacks, Symbol("my-div2.children"))
+    @test app.callbacks[Symbol("my-div.children")].func("value") == "value"
+    @test app.callbacks[Symbol("my-div2.children")].func("value") == "v_value"
+
+    @test_throws ErrorException callback!(app, callid"my-id.value => my-div2.children") do value
```

Comment:
It's not clear to me exactly what's being tested by some of these `@test_throws` - at a minimum it would be nice to have a comment like "my-div2.children is already an output", but better would be for the code to somehow inspect the error to ensure it's actually the one you intended to be thrown, as opposed to something else like a typo in the `callid` string causing a regexp failure.

Minor, and not new in this PR, you're just rearranging, but it's the first I'm looking at this code 😉